### ISSUE TITLE
[$500] Android – Request–Text entered in address input field is not highlighted if click on "Fix error"

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#34048]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
 
**Version Number:** 1.4.22-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Log in
3. Global Create > Request Money
4. Select Distance option
5. Click start
6. Enter any text that will not return an address
7. Try to save
8. Verify an error message displays indicating to select a suggested address or use current location
9. Click on the "Fix error" link in the error displayed in the bottom of the modal

## Expected Result:
The text entered in the address input field is highlighted

## Actual Result:
The text entered in the address input field is not highlighted while click on the "Fix error" link in the error displayed in Distance IOU creation

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/8dfb4b0c-8b5f-42e7-b7b4-92cd5fcd720e

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01af9a1f39eb96a1b2</li>
        <li>Upwork Job ID: 1743338410047598592</li>
        <li>Last Price Increase: 2024-01-05</li>
    </ul>
</details>

**Comment by ikevin127** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

The text entered in the address input field is not highlighted while click on the "Fix error" link in the error displayed in Distance IOU creation.

## What is the root cause of that problem?

Due to platform specific behaviour when an input is focused, on Android the text is not being highlighted by default like it happens on iOS for example.

## What changes do you think we should make in order to solve the problem?

Add `selectTextOnFocus` prop to the input so that the text is always highlighted when the input is focused by pressing the "fix the errors" link.

**Comment by ikevin127** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

The text entered in the address input field is not highlighted while the input is already focused and the Android cursor is showing when pressing the "fix the errors" link of the error displayed in Distance IOU creation.

## What is the root cause of that problem?

The reason why this is happening is due to the fact that the input is already focused and the Android cursor is visible while the "fix the errors" link is pressed, as it can be seen in the OP video.

The "fix the errors" link function can be found [here](https://github.com/Expensify/App/blob/f5a3c0202128a8e0e77607defbbdcb4f30f80ad3/src/components/Form/FormWrapper.js#L125-L151), this function calls focus which has platform specific behaviour:
- on Android if the input is already focused as in OP video, calling focus again does not highlight the text
- on iOS it highlights the whole text even if the input is focused and the iOS cursor is visible

## What changes do you think we should make in order to solve the problem?

My solution would be to add an if that checks whether the input is already focused, and if it is then call blur before the focus. This ensures we get the expected behaviour on Android while not changing the iOS behaviour.

**Comment by ikevin127** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

The text entered in the address input field is not highlighted while the input is already focused and the Android cursor is showing when pressing the "fix the errors" link of the error displayed in Distance IOU creation.

## What is the root cause of that problem?

The reason why this is happening is due to the fact that the input is already focused and the Android cursor is visible while the "fix the errors" link is pressed, as it can be seen in the OP video.

The "fix the errors" link function can be found [here](https://github.com/Expensify/App/blob/f5a3c0202128a8e0e77607defbbdcb4f30f80ad3/src/components/Form/FormWrapper.js#L125-L151), this function calls focus which has platform specific behaviour:
- on Android if the input is already focused as in OP video, calling focus again does not highlight the text
- on iOS it highlights the whole text even if the input is focused and the iOS cursor is visible

## What changes do you think we should make in order to solve the problem?

My solution would be to add an if that checks whether the input is already focused, and if it is then call blur before the focus. This ensures we get the expected behaviour on Android.

The downside is that for a split second the keyboard is being hidden then shown again when focus is called.

### Changed code

```diff
../src/components/Form/FormWrapper.js


onFixTheErrorsLinkPressed={() => {
    ...

+   if (focusInput.isFocused()) {
+       focusInput.blur();
+   }

    // Focus the input after scrolling, as on the Web it gives a slightly better visual result
    if (focusInput.focus && typeof focusInput.focus === 'function') {
        focusInput.focus();
    }
}}

```

## Videos

<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/56457735/ba59eb7c-bb9e-465f-8f60-23f78fd023cc

</details>

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Android – Request–Text entered in address input field is not highlighted if click on "Fix error"
### What is the root cause of that problem?
- Pressing the "Fix error" when the input has been already focused does not trigger the focus event again so the showSoftInputOnFocus does not work in this case.
### What changes do you think we should make in order to solve the problem?
- We can check if the input has the prop showSoftInputOnFocus is true or not and if the input has been already focused or not.
If both conditions are true, call the select all the text input.
- For example, we can update:


### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by s77rt** [comment]
@ikevin127 Thanks for the proposal. Your RCA is correct. However the solution would introduce a flicker which is not something wanted. 

PS: Please do not post code diffs

**Comment by s77rt** [comment]
@DylanDylann Thanks fr the proposal. Your RCA is correct. Regarding the suggested solution

> If both conditions are true, select all the text input

Can you elaborate on this?


PS: Do not raise any PR until assigned. Please explain the solution in plain English

**Comment by s77rt** [comment]
@Krishna2323 Thanks for the proposal. Your RCA is correct. Your suggested solution may only fix this particular case (the distance field). Can we instead work on a global solution (at FormWrapper level)?

**Comment by ikevin127** [comment]
## Updated [proposal](https://github.com/Expensify/App/issues/34048#issuecomment-1879222912)
- updated solution according to requirements from [#34048 (comment)](https://github.com/Expensify/App/issues/34048#issuecomment-1879835838) and added videos for both Android / iOS

**Comment by DylanDylann** [comment]
@s77rt Updated the solution based on your [comment ](https://github.com/Expensify/App/issues/34048#issuecomment-1879834225)

**Comment by s77rt** [comment]
@ikevin127 Your proposal does not take into whether we should select text on focus or not (`selectTextOnFocus` prop)

**Comment by s77rt** [comment]
@DylanDylann Passing all the inputs values and props to `FormWrapper` seems a little overkill. Perhaps we should look for another approach.

**Comment by s77rt** [comment]
@Krishna2323 The `selectTextOnFocus` prop is per input not per form.

**Comment by DylanDylann** [comment]
@s77rt Yeah. I think it is overkill in this case, but in the future, I think we can reuse it for another purpose

**Comment by DylanDylann** [comment]
@s77rt Additionally, we passed the [inputRefs](https://github.com/Expensify/App/blob/0c351a4534f0bfb8278f2fa9b8c0a10187b6c8bb/src/components/Form/FormWrapper.js#L68) to the FormWrapper and just use it in the [onFixTheErrorsLinkPressed](https://github.com/Expensify/App/blob/0c351a4534f0bfb8278f2fa9b8c0a10187b6c8bb/src/components/Form/FormWrapper.js#L125) function

**Comment by ikevin127** [comment]
## Updated [proposal](https://github.com/Expensify/App/issues/34048#issuecomment-1879222912)
- updated solution to account for input's `selectTextOnFocus ` as requested in [#34048 (comment)](https://github.com/Expensify/App/issues/34048#issuecomment-1880124288)

**Comment by s77rt** [comment]
@DylanDylann `focusInput.viewConfig.validAttributes.selectTextOnFocus` will always be `true`. This is telling us that `selectTextOnFocus` is a valid prop and not whether it's passed as true or not.

**Comment by DylanDylann** [comment]
@s77rt Yep. This alternative solution does not work so please ignore it. I forgot to remove it

**Comment by s77rt** [comment]
@ikevin127 Your updated solution is a dupe

**Comment by s77rt** [comment]
@Krishna2323 Your updated solution is also a dupe

**Comment by s77rt** [comment]
Fixing this is not looking worth it given the the complexity / value of the fix ratio. Asking for more opinions on Slack https://expensify.slack.com/archives/C01GTK53T8Q/p1704744305140489

**Comment by s77rt** [comment]
@Krishna2323 I don't see how this is different than just passing the input props/values

**Comment by DylanDylann** [comment]
@s77rt What do you mean about "But the most critical point that concerns me is having to declare selectTextOnFocus twice"?

**Comment by s77rt** [comment]
@Krishna2323 It's not about how much data to send but on what data to send. If we are going to send the input props I prefer to send the whole rather than constructing a new object. Also we should not define selectTextOnFocus as true again because it's already set in AddressSearch. We should read that value from there.

https://github.com/Expensify/App/blob/582b853386af4ac639b35c12f26ab6a8e4bbb7a2/src/components/AddressSearch/index.js#L520

**Comment by s77rt** [comment]
@DylanDylann Because `selectTextOnFocus` prop is not visible to the Form you are suggesting to define it again along the `InputWrapper` which is not right. Now we will have two values that we need to keep coupled (have same value) manually. If we ever change the value on AddressSearch we can forget to change the one we are passing in InputWrapper

**Comment by DylanDylann** [comment]
@s77rt 

- When using the normal TextInput component, we use it like ```<TextInput ...props />```
- When using the TextInput in form, we use it like ```<InputWrapper InputComponent={TextInput} ... props />```
- So, in my demo [PR](https://github.com/Expensify/App/pull/34070), when using the AddressSearch in form, we will use it like 
``` <InputWrapperWithRef InputComponent={AddressSearch} selectTextOnFocus={true} ...props /> ```
So, we just need to pass the ```selectTextOnFocus={true}``` one time

**Comment by DylanDylann** [comment]
> If we ever change the value on AddressSearch

Before, we are hard code the ```selectTextOnInput``` in the AddressSearch:
https://github.com/Expensify/App/blob/f15ed75b0b4133a8c02617d20a09e29d72825a4d/src/components/AddressSearch/index.js#L520
So, my proposal suggested that we should update it to:
```selectTextOnFocus: selectTextOnFocus```
with the selectTextOnFocus is passed from the parent component

**Comment by s77rt** [comment]
@DylanDylann Right. I think I overlooked that. That should do it. Though we will still be passing both props and values. Can you please update your proposal and remove the PR link?

**Comment by DylanDylann** [comment]
@s77rt I think all the solution I `ve mentioned in the proposal so I just need to remove the PR link.

> Can you please update your proposal

- Do you mean this one bellow?

> Note that, we need to pass the prop selectTextOnFocus through the [InputWrapperWithRef](https://github.com/Expensify/App/blob/0c351a4534f0bfb8278f2fa9b8c0a10187b6c8bb/src/pages/iou/request/step/IOURequestStepWaypoint.js#L232) rather than hardcode as we did [here](https://github.com/Expensify/App/blob/0c351a4534f0bfb8278f2fa9b8c0a10187b6c8bb/src/components/AddressSearch/index.js#L520) so the selectTextOnFocus will appear in the inputProps state.

**Comment by s77rt** [comment]
@DylanDylann I have been thinking about this. Let's try do the following:
1. Write a function in FormProvider that accepts an input id and for that input it returns the value and the props
2. Pass that function to FormWrapper
3. In FormWrapper on onFixTheErrorsLinkPressed call that function with the input id having the error

Let me know if we can work on that

**Comment by DylanDylann** [comment]
@s77rt Yes we can do like that

**Comment by s77rt** [comment]
@DylanDylann Can you please update your proposal to go with that

**Comment by DylanDylann** [comment]
@s77rt But I think my original solution (Pass down the values and props to FormWrapper) and you suggestion, that create a function getValueAndProps:
```function getValueAndProps(id) => return {value: inputValues[id], props: inputProps[id]}```
then pass it to FormWrapper are slightly similar, right?

**Comment by s77rt** [comment]
@DylanDylann Right but it's better to update the proposal with what we are going to actually implement so there won't be any mislead when it comes to review.

**Comment by DylanDylann** [comment]
@s77rt I updated the [proposal](https://github.com/Expensify/App/issues/34048#issuecomment-1879241930)

**Comment by s77rt** [comment]
@DylanDylann Thanks. That looks good to me.

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/34048#issuecomment-1879241930)

**Comment by s77rt** [comment]
Not overdue. @aldo-expensify https://github.com/Expensify/App/issues/34048#issuecomment-1888987134

**Comment by aldo-expensify** [comment]
Asked here https://expensify.slack.com/archives/C02NK2DQWUX/p1705329870985479 about this issue because to me it feels of very little value. There is only one input in this form and it already has a big red error, I don't think anyone will miss something because the field did not "highlight".

I find the [proposed solution ](https://github.com/Expensify/App/issues/34048#issuecomment-1879241930) too complex for what we are fixing. I think it is better not to fix it and keep our code simpler.

Sorry for the time wasted in this.
